generator/integration_tests/tests/request_id_connection_impl_test.cc

```diff
@@ -0,0 +1,281 @@
+using ::testing::_;
+using ::testing::ByMove;
+using ::testing::ElementsAre;
+using ::testing::Eq;
+using ::testing::IsEmpty;
+using ::testing::ResultOf;
+using ::testing::Return;
+
+class MockRequestIdServiceStub
+    : public google::cloud::golden_v1_internal::RequestIdServiceStub {
+ public:
+  ~MockRequestIdServiceStub() override = default;
+
+  MOCK_METHOD(StatusOr<google::test::requestid::v1::Foo>, CreateFoo,
+              (grpc::ClientContext&,
+               google::test::requestid::v1::CreateFooRequest const&),
+              (override));
+
+  MOCK_METHOD(future<StatusOr<google::longrunning::Operation>>, AsyncRenameFoo,
+              (google::cloud::CompletionQueue&,
+               std::shared_ptr<grpc::ClientContext>, Options const&,
+               google::test::requestid::v1::RenameFooRequest const&),
+              (override));
+
+  MOCK_METHOD(StatusOr<google::test::requestid::v1::ListFoosResponse>, ListFoos,
+              (grpc::ClientContext&,
+               google::test::requestid::v1::ListFoosRequest const&),
+              (override));
+
+  MOCK_METHOD(future<StatusOr<google::test::requestid::v1::Foo>>,
+              AsyncCreateFoo,
+              (google::cloud::CompletionQueue&,
+               std::shared_ptr<grpc::ClientContext>,
+               google::test::requestid::v1::CreateFooRequest const&),
+              (override));
+
+  MOCK_METHOD(future<StatusOr<google::longrunning::Operation>>,
+              AsyncGetOperation,
+              (google::cloud::CompletionQueue&,
+               std::shared_ptr<grpc::ClientContext>, Options const&,
+               google::longrunning::GetOperationRequest const&),
+              (override));
+
+  MOCK_METHOD(future<Status>, AsyncCancelOperation,
+              (google::cloud::CompletionQueue&,
+               std::shared_ptr<grpc::ClientContext>, Options const&,
+               google::longrunning::CancelOperationRequest const&),
+              (override));
+};
+
+template <typename Request>
+auto WithRequestId(std::string expected) {
+  return ResultOf(
+      "request has request_id",
+      [](Request const& request) { return request.request_id(); },
+      Eq(std::move(expected)));
+}
+
+template <typename Request>
+auto WithRequestId() {
+  return ResultOf(
+      "request does not have request_id",
+      [](Request const& request) { return request.request_id(); }, IsEmpty());
+}
+
+auto MakeTestConnection(
+    std::shared_ptr<golden_v1_internal::RequestIdServiceStub> stub) {
+  auto options = golden_v1_internal::RequestIdServiceDefaultOptions({});
+  auto background = internal::MakeBackgroundThreadsFactory(options)();
+  return std::make_shared<golden_v1_internal::RequestIdServiceConnectionImpl>(
+      std::move(background), std::move(stub), std::move(options));
+}
+
+TEST(RequestIdTest, UnaryRpc) {
+  auto mock = std::make_shared<MockRequestIdServiceStub>();
+  EXPECT_CALL(*mock, CreateFoo(_, WithRequestId<CreateFooRequest>()))
+      .WillOnce(Return(Foo{}));
```

Comment:
Should we return UNAVAILABLE to check that retries use the proper request ID?

Reply:
Good idea, once these things are actually working.